usr.bin/cvs/cvs.1
.\"	$OpenBSD: cvs.1,v 1.45 2005/02/09 16:11:15 xsa Exp $
.\"
.\" Copyright (c) 2004 Jean-Francois Brousseau <jfb@openbsd.org>
.\" Copyright (c) 2004, 2005 Xavier Santolaria <xsa@openbsd.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
.\" INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
.\" AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
.\" THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
.\" EXEMPLARY, OR CONSEQUENTIAL  DAMAGES (INCLUDING, BUT NOT LIMITED TO,
.\" PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
.\" OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
.\" WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
.\" OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
.\" ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd May 16, 2004
.Dt CVS 1
.Os
.Sh NAME
.Nm cvs
.Nd OpenCVS Concurrent Versioning System client
.Sh SYNOPSIS
.Nm
.Op Fl flQqtv
.Op Fl d Ar root
.Op Fl e Ar editor
.Xo
.Oo Fl s
.Ar var Ns = Ns Ar val Oc
.Xc
.Op Fl z Ar level
.Ar command Op Ar ...
.Sh DESCRIPTION
The
.Nm
program acts as both client and server for the use of and administration of
a CVS source repository.
CVS is used to maintain version information on files that are kept in a
repository.
Although it is more commonly used to track changes in source code, there
are no real limitations to the type of files that you can store in a
repository.
For a general introduction to CVS, see
.Xr cvsintro 7 .
.Pp
.Nm
reads its startup configuration file,
.Pa .cvsrc ,
from the home directory of the user who invoked it.
This file is used to specify implicit options passed to
.Nm
or one of its commands whenever it is invoked.
The defaults in the configuration file can be overridden with the
.Fl f
option (see below).
See
.Xr cvsrc 5
for further information.
.Pp
The following options are supported:
.Bl -tag -width "-e editorXX"
.It Fl d Ar root
Use
.Ar root
as the path to the root directory of the CVS repository.
The value must specify an absolute path.
.It Fl e Ar editor
Use the program
.Ar editor
whenever editing log information.
This option overrides the environment variables CVSEDITOR, VISUAL, and EDITOR.
.It Fl f
Do not read the user's configuration file on startup.
.It Fl l
Suppress logging of history information.
.It Fl Q
Be extra quiet.
Only error messages will be displayed.
.It Fl q
Be quiet about reporting.
.It Fl s Ar var Ns = Ns Ar val
Set the value of the internal variable
.Ar var
to the string
.Ar val .
.It Fl t
Trace program execution.
.It Fl v
Display version information and exit.
.It Fl z Ar level
Specify the compression level to
.Xr gzip 1
when transferring files.
The compression level ranges from 1 to 9,
with 1 being the fastest,
and 9 providing the best level of compression.
The default is 6.
.El
.Sh KEYWORD SUBSTITUTION
As long as you edit source files inside a working directory you
can always find out the state of your files via the
.Ic status
or
.Ic log
commands, but as soon as files get exported from
your local working copy, it becomes harder to identify which
revisions they are.
.Pp
.Nm
can use a mechanism known as
.Sq keyword substitution
to help identify the files.
Embedded strings of the form $keyword$ and $keyword:...$ in a file
are replaced with strings of the form $keyword: value$ whenever you
obtain a new revision of the file.
The possible keywords are as follows:
.Bl -tag -width "RevisionXXX"
.It $Author: xsa $
The name of the user who checked in the revision.
.It $Date: 2005/02/09 16:11:15 $
The date and hour (UTC) the revision was checked in.
.It $Header: /scm/cvs/src/usr.bin/cvs/cvs.1,v 1.45 2005/02/09 16:11:15 xsa Exp $
Standard header containing the full pathname of the RCS
file, the revision number, the date (UTC), the author and the state.
.It $Id: cvs.1,v 1.45 2005/02/09 16:11:15 xsa Exp $
The same content as $Header: /scm/cvs/src/usr.bin/cvs/cvs.1,v 1.45 2005/02/09 16:11:15 xsa Exp $ but without the path
of the RCS file.
.It $Log: cvs.1,v $
.It Revision 1.45  2005/02/09 16:11:15  xsa
.It add a KEYWORD SUBSTITUTION section; ok jmc@ jfb@ Joris Vink;
.It
The log message supplied during commit, preceded by a header
containing the RCS filename, the revision number, the
author, and the date (UTC).
.It $Name:  $
The tag name used to check out the file.
.It $RCSfile: cvs.1,v $
The name of the RCS file, but without a path.
.It $Revision: 1.45 $
The revision number assigned to the revision.
.It $Source: /scm/cvs/src/usr.bin/cvs/cvs.1,v $
The full pathname of the RCS file.
.It $State: Exp $
The state assigned to the revision.
.El
.Pp
Keyword substitution has its disadvantages: sometimes the
literal text string $Author: xsa $ is wanted inside a file without
.Nm
interpreting it as a keyword and expanding it into something like
$Author: xsa $.
The
.Fl k Ar o
option can be used to turn off keyword substitution entirely though. 
There is unfortunately no way to selectively turn off keyword substitution.
.Pp
Each file and working directory copy of a file have a stored
default substitution mode.
Substitution modes on files are set by the
.Fl k Ar mode
option to the
.Ic add
and
.Ic admin
commands or by the
.Fl k Ar mode
or
.Fl A
options to the
.Ic checkout
or
.Ic update
commands.
.Pp
The possible substitution modes are as follows:
.Bl -tag -width Ds -offset 3n
.It Fl k Ar b
Like
.Fl k Ar o ,
but also avoids the conversion of line endings.
This option is used to handle binary files.
.It Fl k Ar k
Does not substitute the keywords.
Useful with the
.Ic diff
command to avoid displaying the differences between keyword substitutions.
.It Fl k Ar kv
The default behaviour.
Keywords are normally substituted i.e. $Revision: 1.45 $ becomes $Revision: 1.45 $.
.It Fl k Ar kvl
Like
.Fl k Ar kv ,
except that the locker's name is displayed along with the version
if the given revision is currently locked.
This option is normally not useful as
.Nm
does not use file locking by default.
.It Fl k Ar o
No substitutions are done.
This option is often used with the
.Ic import
command to guarantee that files that already contain external keywords
do not get modified.
.It Fl k Ar v
Substitue the value of keywords instead of keywords themselves
e.g. instead of $Revision: 1.45 $, only insert 1.1 and not $Revision: 1.45 $.
This option must be used with care, as it can only be used once.
It is often used with the
.Ic export
command to freeze the values before releasing software.
.El
.Sh COMMANDS
The following commands are supported by
.Nm :
.Bl -tag -width "xxxxxxxxxxxx"
.It Xo Ic add
.Op Fl k Ar mode
.Op Fl m Ar msg
.Ar file ...
.Xc
.Pp
Before a file is known to
.Nm ,
it must be added to the repository using this command.
Adding a file does not actually publish the contents of the
file: the
.Ic commit
command must also be used to publish it into the repository,
and thus let others access the file.
.Pp
Note: since directories have no versioning system, it is sufficient
to add them with the
.Ic add
command alone; the
.Ic commit
command is not necessary.
.Pp
The
.Ic add
command takes the following options:
.Bl -tag -width Ds -offset 3n
.It Fl k Ar mode
Specify the keyword substitution mode.
.It Fl m Ar msg
Attach log message
.Ar msg .
By default, no log message is required.
.El
.Pp
Aliases:
.Ic ad ,
.Ic new .
.It Ic admin
Administration front-end for
.Xr rcs 1 .
.Pp
Aliases:
.Ic adm ,
.Ic rcs .
.It Xo Ic annotate
.Op Fl flR
.Oo Fl D Ar date \*(Ba
.Fl r Ar rev Oc
.Op Ar file ...
.Xc
.Pp
For each line of any files specified, show information about its
last revision.
The information given is the last revision when a modification occurred,
the author's name, and the date of the revision.
.Pp
The
.Ic annotate
command takes the following options:
.Bl -tag -width Ds -offset 3n
.It Fl D Ar date
Show the annotations as of the latest revision no later than
.Ar date .
.It Fl f
Force the use of the head revision if the specified
tag or date is not found.
This can be used in combination with
.Fl D
or
.Fl r
to ensure that there is some output from the
.Ic annotate
command, even if only to show Revision 1.1 of the file.
.It Fl l
Limit the scope of the search to the local directory
only and disable recursive behaviour.
.It Fl R
Enable recursive behaviour.
This is the default.
.It Fl r Ar rev
Show annotations as of revision
.Ar rev
(can be a revision number or a tag).
.El
.Pp
Aliases:
.Ic ann .
.It Xo Ic checkout
.Op Fl AcflNnPpRs
.Op Fl d Ar dir
.Op Fl j Ar rev
.Op Fl k Ar mode
.Fl D Ar date \*(Ba
.Fl r Ar rev
.Ar module ...
.Xc
.Pp
The
.Ic checkout
command is used to create a local copy of one or more modules present on the
target CVS repository.
.Pp
The
.Ic checkout
command takes the following options:
.Bl -tag -width Ds -offset 3n
.It Fl A
Reset any sticky tags, dates, or keyword substitution modes that
have been set on the tree.
.It Fl c
Display the list of available modules.
.It Fl D Ar date
Check out as of the latest revision no later than
.Ar date
(is sticky).
.It Fl d Ar dir
Check out in directory
.Ar dir
instead of the directory bearing the same name as the
.Ar module .
.It Fl f
Force the use of the head revision if the specified
tag or date is not found.
.It Fl j Ar rev
Merge in changes made between current revision and
.Ar rev .
If two
.Fl j
options are specified, only merge the differences between the two
revisions of the branch.
This allows successive merges without having to resolve
already resolved conflicts again.
.It Fl k Ar mode
Specify the keyword substitution mode (is sticky).
.It Fl l
Limit the scope of the search to the local directory
only and disable recursive behaviour.
.It Fl N
If used in conjunction with the
.Fl d
option, files are placed in local directory
.Ar module ,
located in directory
.Ar dir .
.It Fl n
Do not execute programs listed in the
.Pa CVSROOT/modules
file.
.It Fl P
Prune empty directories.
.It Fl p
Check out files to standard output (avoids stickiness).
.It Fl R
Enable recursive behaviour.
This is the default.
.It Fl r Ar rev
Check out from a particular revision or branch (implies
.Fl P )
(is sticky).
.It Fl s
Like
.Fl c ,
but include module status.
.El
.Pp
Aliases:
.Ic co ,
.Ic get .
.It Xo Ic commit
.Op Fl flnR
.Oo Fl F Ar logfile \*(Ba
.Fl m Ar msg Oc
.Op Fl r Ar rev
.Op Ar file ...
.Xc
.Pp
The
.Ic commit
command is used to send local changes back to the server and update the
repository's information to reflect the changes.
.Pp
The
.Ic commit
command takes the following options:
.Bl -tag -width Ds -offset 3n
.It Fl F Ar logfile
Specify a
.Ar file
which contains the log message.
.It Fl f
Force a file to be committed, even though it is unchanged.
.It Fl l
Limit the scope of the search to the local directory
only and disable recursive behaviour.
.It Fl m Ar msg
Specify a log message on the command line (suppresses the editor invocation).
.It Fl n
Do not execute programs listed in the
.Pa CVSROOT/modules
file.
.It Fl R
Enable recursive behaviour.
This is the default.
.It Fl r Ar rev
Commit to a particular symbolic or numerical revision.
.El
.Pp
Aliases:
.Ic ci ,
.Ic com .
.It Xo Ic diff Op Fl cilNpu
.Op Fl D Ar date
.Op Fl r Ar rev
.Op Ar file ...
.Xc
.Pp
The
.Ic diff
command is very similar to the
.Xr diff 1
program, except that the differential comparisons that it generates are
between local or remote revisions of files stored in the CVS repository.
.Pp
Aliases:
.Ic di ,
.Ic dif .
.It Xo Ic edit Op Fl lR
.Op Fl a Ar action
.Op Ar file ...
.Xc
.Pp
The
.Ic edit
command is used to make a file that is being watched
(and therefore read-only)
readable and writable and to inform others that you are planning to edit it.
Notifications terminate when the
.Ic commit
command is issued.
Editing rights on the file can be given up using the
.Ic unedit
command, which terminates the temporary notifications.
.Pp
The
.Ic edit
command takes the following options:
.Bl -tag -width Ds -offset 3n
.It Fl a Ar action
Specify the temporary notification wanted:
.Pp
.Bl -tag -width "commitXX" -compact
.It Cm commit
Another user has committed changes to the file.
.It Cm edit
Another user has issued the
.Ic edit
command on the file.
.It Cm unedit
Another user has issued the
.Ic unedit
command on the file.
.It Cm all
All of the above.
.It Cm none
None of the above.
.El
.Pp
The
.Fl a
flag may appear more than once, or not at all.
If omitted, the action defaults to
.Cm all .
.It Fl l
Limit the scope of the search to the local directory
only and disable recursive behaviour.
.It Fl R
Enable recursive behaviour.
This is the default.
.El
.It Xo Ic editors
.Op Fl lR
.Op Ar file ...
.Xc
.Pp
The
.Ic editors
command lists the users with edition rights on a file.
For that, pseudo-lock mode must be enabled (see the
.Ic watch
command).
The e-mail address of the user editing the file, the timestamp
when the edition first started, the host from where the edition
has been requested and the path to the edited file are listed.
.Pp
The
.Ic editors
command takes the following options:
.Bl -tag -width Ds -offset 3n
.It Fl l
Limit the scope of the search to the local directory
only and disable recursive behaviour.
.It Fl R
Enable recursive behaviour.
This is the default.
.El
.It Xo Ic export
.Op Fl flNnR
.Op Fl d Ar dir
.Op Fl k Ar mode
.Fl D Ar date \*(Ba
.Fl r Ar rev
.Ar module ...
.Xc
.Pp
The
.Ic export
command extracts a copy of
.Ar module
without including the directories used for management by
.Nm .
This eases production of a software release.
A date or a revision must be specified for the command to be valid,
which ensures that later extractions can be reproduced with the same
options as the release.
.Pp
The checked out module's files will be placed in a directory
bearing the same name as the checked out module, by default.
.Pp
The
.Ic export
command takes the following options:
.Bl -tag -width Ds -offset 3n
.It Fl D Ar date
Export as of the latest revision no later than
.Ar date .
.It Fl d Ar dir
Export in directory
.Ar dir
instead of the directory bearing the same name as the
.Ar module .
.It Fl f
Force the use of the head revision if the specified
tag or date is not found.
This can be used in combination with
.Fl D
or
.Fl r
to ensure that the
.Ic export
command is valid.
.It Fl k Ar mode
Specify the keyword substitution mode: the
.Fl k Ar v
option is often used to avoid substitution of keywords during
a release cycle.
However, be aware that it does not handle an export containing
binary files correctly.
.It Fl l
Limit the scope of the search to the local directory
only and disable recursive behaviour.
.It Fl N
If used in conjunction with the
.Fl d
option, files are placed in local directory
.Ar module ,
located in directory
.Ar dir .
.It Fl n
Do not execute programs listed in the
.Pa CVSROOT/modules
file.
.It Fl R
Enable recursive behaviour.
This is the default.
.It Fl r Ar rev
Export from a particular symbolic or numerical revision.
.El
.Pp
Aliases:
.Ic ex ,
.Ic exp .
.It Ic history
Show repository access history.
.Pp
Aliases:
.Ic hi ,
.Ic his .
.It Xo Ic import
.Op Fl b Ar branch
.Op Fl m Ar msg
.Ar repository
.Ar vendortag
.Ar releasetag
.Xc
.Pp
Import sources into CVS using vendor branches.
.Pp
At least three arguments are required:
.Ar repository
specifies the location of the sources to be imported;
.Ar vendortag
is a tag for the entire branch;
.Ar releasetag
is used to identify the files you created with
.Ic cvs import .
.Pp
The
.Ic import
command takes the following options:
.Bl -tag -width Ds -offset 3n
.It Fl b Ar branch
Specify the first-level branch number.
.It Fl m Ar msg
Specify the log message to send.
.El
.Pp
Aliases:
.Ic im ,
.Ic imp .
.It Ic init
Create a CVS repository if it doesn't exist.
.It Ic kserver
Start a Kerberos authentication server.
.It Ic log
Print out history information for files.
.Pp
Aliases:
.Ic lo .
.It Ic login
Prompt for a password for an authenticating server.
.It Ic logout
Remove an entry in
.Pa .cvspass
for a remote repository.
.It Ic rdiff
Create
.Xr patch 1
format diffs between releases.
.It Xo Ic release
.Op Fl d
.Ar dir ...
.Xc
The
.Ic release
command indicates to
.Nm
that the working copy of a module is no longer in use and checks
that non archived modifications in the base repository do exist.
This command is not mandatory.
Local directories could always be removed without using it, but
in this case the handling of history information will no longer be
correct (see the
.Ic history
command).
.Pp
The
.Ic release
command takes the following options:
.Bl -tag -width Ds -offset 3n
.It Fl d Ar dir
Remove the directory
.Ar dir .
Be aware that this option silently removes any directories that have
been added to the local working copy without using the
.Ic add
command.
.El
.Pp
For each file not being synchronized with the base repository,
a single letter prefix is given to specify the state of the file.
The possible prefixes are as follows:
.Bl -tag -width "XXX"
.It \&?
The file is unknown to
.Nm
and is not in the list of files to ignore.
Any new directories which have not been added with the
.Ic add
command are silently ignored as well as their content.
.It A
The file has been added with the
.Ic add
command, but has not been committed to the repository with the
.Ic commit
command.
.It M
The file has been locally modified; a more recent version might
exist in the base repository.
.It R
The file has been removed with the
.Ic remove
command, but has not been committed to the repository with the
.Ic commit
command.
.It U
A more recent version of the file does exist but it is not
locally up to date.
.El
.Pp
Aliases:
.Ic re ,
.Ic rel .
.It Xo Ic remove
.Op Fl flR
.Op Ar file ...
.Xc
The
.Ic remove
command is used to inform
.Nm
that
.Ar file
is scheduled to be removed from the repository.
Files are not actually removed from the repository until the
.Ic commit
command has been run subsequently.
.Pp
The
.Ic remove
command takes the following options:
.Bl -tag -width Ds -offset 3n
.It Fl f
Force local file removal.
If this flag is not used, the file must be locally removed beforehand for
the command to be valid.
.It Fl l
Limit the scope of the search to the local directory
only and disable recursive behaviour.
.It Fl R
Enable recursive behaviour.
This is the default.
.El
.Pp
Aliases:
.Ic rm ,
.Ic delete .
.It Ic rlog
Print out history information for a module.
.It Ic rtag
Add a symbolic link to a module.
.It Ic server
Server mode.
.It Xo Ic status
.Op Fl lRv
.Op Ar file ...
.Xc
The
.Ic status
command is used to display the state of checked out files.
.Pp
The
.Ic status
command takes the following options:
.Bl -tag -width Ds -offset 3n
.It Fl l
Limit the scope of the search to the local directory
only and disable recursive behaviour.
.It Fl R
Enable recursive behaviour.
This is the default.
.It Fl v
Display symbolic tags for
.Ar file .
.Pp
The state may be one of the following:
.Bl -tag -width "Locally modified"
.It Cm Locally Added
The file has been added with the
.Ic add
command, but has not been committed to the repository with the
.Ic commit
command.
.It Cm Locally Modified
The file is up to date, but has been locally modified.
.It Cm Locally Removed
The file has been removed with the
.Ic remove
command, but has not been committed to the repository with the
.Ic commit
command.
.It Cm Needs Checkout
The file has not been modified; a new version is available.
.It Cm Needs Merge
The file has been modified and a newer version is available.
.It Cm Needs Patch
Same as
.Ic Needs Checkout
but, in client-server mode, only the differences are sent to save
network resources.
.It Cm Unresolved Conflict
A merge has been done, but unresolved conflicts still remain.
.It Cm Up-to-date
The file is up to date.
.El
.El
.Pp
Aliases:
.Ic st ,
.Ic stat .
.It Xo Ic tag
.Op Fl bdlR
.Op Fl r Ar old_tag
.Op Ar symbolic_tag
.Op Ar file ...
.Xc
.Pp
Add a symbolic tag to a checked out version of
.Ar file .
.Pp
The
.Ic tag
command takes the following options:
.Bl -tag -width Ds -offset 3n
.It Fl b
Specify a branch tag.
.It Fl d
Delete tag.
.It Fl l
Limit the scope of the search to the local directory
only and disable recursive behaviour.
.It Fl R
Enable recursive behaviour.
This is the default.
.It Fl r Ar old_tag
.Nm
will only tag the files with
.Ar symbolic_tag
if they are already tagged with
.Ar old_tag .
When done, it will remove the old tag
leaving only the new tag behind on exactly the same files.
.El
.Pp
Aliases:
.Ic ta ,
.Ic freeze .
.It Xo Ic unedit Op Fl lR
.Op Ar file ...
.Xc
.Pp
The
.Ic unedit
command is used to give up an edition on a file and thus cancel
the wanted temporary notifications.
If the file has been modified since the
.Ic edit
command has been issued,
.Nm
will ask if you want to go back to the previous version, and lose the
modifications done on the file, or stay in edition mode on it.
.Pp
The
.Ic unedit
command takes the following options:
.Bl -tag -width Ds -offset 3n
.It Fl l
Limit the scope of the search to the local directory
only and disable recursive behaviour.
.It Fl R
Enable recursive behaviour.
This is the default.
.El
.It Xo Ic update
.Op Fl AdflPpR
.Oo Fl D Ar date \*(Ba
.Fl r Ar rev Oc
.Op Fl I Ar ign
.Op Fl j Ar rev

usr.bin/cvs/cvs.c
#define DEBUG
/*	$OpenBSD: cvs.c,v 1.1 2004/07/13 22:02:40 jfb Exp $	*/
/*
 * Copyright (c) 2004 Jean-Francois Brousseau <jfb@openbsd.org>
 * All rights reserved. 
 *
 * Redistribution and use in source and binary forms, with or without 
 * modification, are permitted provided that the following conditions 
 * are met: 
 *
 * 1. Redistributions of source code must retain the above copyright 
 *    notice, this list of conditions and the following disclaimer. 
 * 2. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission. 
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
 * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
 * AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
 * THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL  DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
 */

#include <sys/types.h>
#include <sys/wait.h>

#include <err.h>
#include <pwd.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <string.h>
#include <sysexits.h>

#include "cvs.h"
#include "log.h"


extern char *__progname;


/* verbosity level: 0 = really quiet, 1 = quiet, 2 = verbose */
int verbosity = 2;



/* compression level used with zlib, 0 meaning no compression taking place */
int   cvs_compress = 0;
int   cvs_trace = 0;
int   cvs_nolog = 0;
int   cvs_readonly = 0;

/* name of the command we are running */
char *cvs_command;
char *cvs_rootstr;
char *cvs_rsh = CVS_RSH_DEFAULT;
char *cvs_editor = CVS_EDITOR_DEFAULT;

struct cvsroot *cvs_root = NULL;


/*
 * Command dispatch table
 * ----------------------
 *
 * The synopsis field should only contain the list of arguments that the
 * command supports, without the actual command's name.
 *
 * Command handlers are expected to return 0 if no error occured, or one of
 * the values known in sysexits.h in case of an error.  In case the error
 * returned is EX_USAGE, the command's usage string is printed to standard
 * error before returning.
 */

static struct cvs_cmd {
	char    cmd_name[CVS_CMD_MAXNAMELEN];
	char    cmd_alias[CVS_CMD_MAXALIAS][CVS_CMD_MAXNAMELEN];
	int   (*cmd_hdlr)(int, char **);
	char   *cmd_synopsis;
	char    cmd_descr[CVS_CMD_MAXDESCRLEN];
} cvs_cdt[] = {
	{
		"add",      { "ad",  "new" }, cvs_add,
		"[-m msg] file ...",
		"Add a new file/directory to the repository",
	},
	{
		"admin",    { "adm", "rcs" }, NULL,
		"",
		"Administration front end for rcs",
	},
	{
		"annotate", { "ann"        }, NULL,
		"",
		"Show last revision where each line was modified",
	},
	{
		"checkout", { "co",  "get" }, NULL,
		"",
		"Checkout sources for editing",
	},
	{
		"commit",   { "ci",  "com" }, cvs_commit,
		"",
		"Check files into the repository",
	},
	{
		"diff",     { "di",  "dif" }, cvs_diff,
		"",
		"Show differences between revisions",
	},
	{
		"edit",     {              }, NULL,
		"",
		"Get ready to edit a watched file",
	},
	{
		"editors",  {              }, NULL,
		"",
		"See who is editing a watched file",
	},
	{
		"export",   { "ex",  "exp" }, NULL,
		"",
		"Export sources from CVS, similar to checkout",
	},
	{
		"history",  { "hi",  "his" }, cvs_history,
		"",
		"Show repository access history",
	},
	{
		"import",   { "im",  "imp" }, NULL,
		"",
		"Import sources into CVS, using vendor branches",
	},
	{
		"init",     {              }, cvs_init,
		"",
		"Create a CVS repository if it doesn't exist",
	},
#if defined(HAVE_KERBEROS)
	{
		"kserver",  {}, NULL
		"",
		"Start a Kerberos authentication CVS server",
	},
#endif
	{
		"log",      { "lo"         }, cvs_getlog,
		"",
		"Print out history information for files",
	},
	{
		"login",    {}, NULL,
		"",
		"Prompt for password for authenticating server",
	},
	{
		"logout",   {}, NULL,
		"",
		"Removes entry in .cvspass for remote repository",
	},
	{
		"rdiff",    {}, NULL,
		"",
		"Create 'patch' format diffs between releases",
	},
	{
		"release",  {}, NULL,
		"",
		"Indicate that a Module is no longer in use",
	},
	{
		"remove",   {}, NULL,
		"",
		"Remove an entry from the repository",
	},
	{
		"rlog",     {}, NULL,
		"",
		"Print out history information for a module",
	},
	{
		"rtag",     {}, NULL,
		"",
		"Add a symbolic tag to a module",
	},
	{
		"server",   {}, cvs_server,
		"",
		"Server mode",
	},
	{
		"status",   {}, NULL,
		"",
		"Display status information on checked out files",
	},
	{
		"tag",      { "ta", }, NULL,
		"",
		"Add a symbolic tag to checked out version of files",
	},
	{
		"unedit",   {}, NULL,
		"",
		"Undo an edit command",
	},
	{
		"update",   {}, cvs_update,
		"",
		"Bring work tree in sync with repository",
	},
	{
		"version",  {}, cvs_version,
		"",
		"Show current CVS version(s)",
	},
	{
		"watch",    {}, NULL,
		"",
		"Set watches",
	},
	{
		"watchers", {}, NULL,
		"",
		"See who is watching a file",
	},
};

#define CVS_NBCMD  (sizeof(cvs_cdt)/sizeof(cvs_cdt[0]))



void             usage        (void);
void             sigchld_hdlr (int);
void             cvs_readrc   (void);
struct cvs_cmd*  cvs_findcmd  (const char *); 



/*
 * sigchld_hdlr()
 *
 * Handler for the SIGCHLD signal, which can be received in case we are
 * running a remote server and it dies.
 */

void
sigchld_hdlr(int signo)
{
	int status;
	pid_t pid;

	if ((pid = wait(&status)) == -1) {
	}
}


/*
 * usage()
 *
 * Display usage information.
 */

void
usage(void)
{
	fprintf(stderr,
	    "Usage: %s [-lQqtv] [-d root] [-e editor] [-z level] "
	    "command [options] ...\n",
	    __progname);
}


int
main(int argc, char **argv)
{
	char *envstr, *ep;
	int ret;
	u_int i, readrc;
	struct cvs_cmd *cmdp;

	readrc = 1;

	if (cvs_log_init(LD_STD, 0) < 0)
		err(1, "failed to initialize logging");

	/* by default, be very verbose */
	(void)cvs_log_filter(LP_FILTER_UNSET, LP_INFO);

#ifdef DEBUG
	(void)cvs_log_filter(LP_FILTER_UNSET, LP_DEBUG);
#endif

	/* check environment so command-line options override it */
	if ((envstr = getenv("CVS_RSH")) != NULL)
		cvs_rsh = envstr;

	if (((envstr = getenv("CVSEDITOR")) != NULL) ||
	    ((envstr = getenv("VISUAL")) != NULL) ||
	    ((envstr = getenv("EDITOR")) != NULL))
		cvs_editor = envstr;

	while ((ret = getopt(argc, argv, "d:e:fHlnQqrtvz:")) != -1) {
		switch (ret) {
		case 'd':
			cvs_rootstr = optarg;
			break;
		case 'e':
			cvs_editor = optarg;
			break;
		case 'f':
			readrc = 0;
			break;
		case 'l':
			cvs_nolog = 1;
			break;
		case 'n':
			break;
		case 'Q':
			verbosity = 0;
			break;
		case 'q':
			/* don't override -Q */
			if (verbosity > 1)
				verbosity = 1;
			break;
		case 'r':
			cvs_readonly = 1;
			break;
		case 't':
			cvs_trace = 1;
			break;
		case 'v':
			printf("%s\n", CVS_VERSION);
			exit(0);
			/* NOTREACHED */
			break;
		case 'z':
			cvs_compress = (int)strtol(optarg, &ep, 10); 
			if (*ep != '\0')
				errx(1, "error parsing compression level");
			if (cvs_compress < 0 || cvs_compress > 9)
				errx(1, "gzip compression level must be "
				    "between 0 and 9");
			break;
		default:
			usage();
			exit(EX_USAGE);
		}
	}

	argc -= optind;
	argv += optind;

	/* reset getopt() for use by commands */
	optind = 1;
	optreset = 1;

	if (argc == 0) {
		usage();
		exit(EX_USAGE);
	}

	/* setup signal handlers */
	signal(SIGCHLD, sigchld_hdlr);

	if (readrc)
		cvs_readrc();

	cvs_command = argv[0];
	ret = -1;

	cmdp = cvs_findcmd(cvs_command);
	if (cmdp == NULL) {
		fprintf(stderr, "Unknown command: `%s'\n\n", cvs_command);
		fprintf(stderr, "CVS commands are:\n");
		for (i = 0; i < CVS_NBCMD; i++)
			fprintf(stderr, "\t%-16s%s\n",
			    cvs_cdt[i].cmd_name, cvs_cdt[i].cmd_descr);
		exit(EX_USAGE);
	}

	if (cmdp->cmd_hdlr == NULL) {
		cvs_log(LP_ERR, "command `%s' not implemented", cvs_command);
		exit(1);
	}

	ret = (*cmdp->cmd_hdlr)(argc, argv);
	if (ret == EX_USAGE) {
		fprintf(stderr, "Usage: %s %s %s\n", __progname, cvs_command,
		    cmdp->cmd_synopsis);
	}

	return (ret);
}


/*
 * cvs_findcmd()
 *
 * Find the entry in the command dispatch table whose name or one of its
 * aliases matches <cmd>.
 * Returns a pointer to the command entry on success, NULL on failure.
 */

struct cvs_cmd*
cvs_findcmd(const char *cmd)
{
	u_int i, j;
	struct cvs_cmd *cmdp;

	cmdp = NULL;

	for (i = 0; (i < CVS_NBCMD) && (cmdp == NULL); i++) {
		if (strcmp(cmd, cvs_cdt[i].cmd_name) == 0)
			cmdp = &cvs_cdt[i];
		else {
			for (j = 0; j < CVS_CMD_MAXALIAS; j++) {
				if (strcmp(cmd, cvs_cdt[i].cmd_alias[j]) == 0) {
					cmdp = &cvs_cdt[i];
					break;
				}
			}
		}
	}

	return (cmdp);
}


/*
 * cvs_readrc()
 *
 * Read the CVS `.cvsrc' file in the user's home directory.  If the file
 * exists, it should contain a list of arguments that should always be given
 * implicitly to the specified commands.
 */

void
cvs_readrc(void)
{
	char rcpath[MAXPATHLEN], linebuf[128], *lp;
	struct cvs_cmd *cmdp;
	struct passwd *pw;
	FILE *fp;

	pw = getpwuid(getuid());
	if (pw == NULL) {
		cvs_log(LP_NOTICE, "failed to get user's password entry");
		return;
	}

	snprintf(rcpath, sizeof(rcpath), "%s/%s", pw->pw_dir, CVS_PATH_RC);

	fp = fopen(rcpath, "r");
	if (fp == NULL) {
		if (errno != ENOENT)
			cvs_log(LP_NOTICE, "failed to open `%s': %s", rcpath,
			    strerror(errno));
		return;
	}

	while (fgets(linebuf, sizeof(linebuf), fp) != NULL) {
		lp = strchr(linebuf, ' ');

		/* ignore lines with no arguments */
		if (lp == NULL)
			continue;

		*(lp++) = '\0';
		if (strcmp(linebuf, "cvs") == 0) {
			/* global options */
		}
		else {
			cmdp = cvs_findcmd(linebuf);
			if (cmdp == NULL) {
				cvs_log(LP_NOTICE,
				    "unknown command `%s' in cvsrc",
				    linebuf);
				continue;
			}
		}
	}
	if (ferror(fp)) {
		cvs_log(LP_NOTICE, "failed to read line from cvsrc");
	}

	(void)fclose(fp);
}
